sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashMapGenerator.scala

```diff
@@ -173,7 +173,10 @@ abstract class HashMapGenerator(
             ${hashBytes(bytes)}
           """
         }
-      case _: StringType => hashBytes(s"$input.getBytes()")
+      case StringType => hashBytes(s"$input.getBytes()")
+      case st: StringType =>
+        hashLong(s"CollationFactory.fetchCollation(${st.collationId})" +
```

Comment:
is this needed by this PR?

Reply:
It is. The thing is that we are still going to use HashAgg for binary collations (e.g. UNICODE_CI which is also covered by tests). We need to do proper hashing in this case.

Reply:
I see. I feel the code can be more readable if we do
```
case st: StringType =>
  if (st.isDefaultCollation) {

  } else {

  }
```

Reply:
I did some refactoring here. Please take a look. I kept pattern matching but I think it looks a bit nicer now.